raft/log.go

```diff
@@ -586,289 +628,359 @@ func (l *Log) Leave() error {
 	return nil
 }
 
-// setState moves the log to a given state.
-func (l *Log) setState(state State) {
-	l.Logger.Printf("log state change: %s => %s", l.state, state)
+// startStateLoop begins the state loop in a separate goroutine.
+// Returns once the state has transitioned from "stopped".
+func (l *Log) startStateLoop(closing <-chan struct{}, state State) {
+	l.wg.Add(1)
+	go l.stateLoop(closing, state)
 
-	// Stop previous state.
-	if l.ch != nil {
-		close(l.ch)
-		l.ch = nil
+	// Wait until state change.
+	for {
+		if l.state == state {
+			break
+		}
+		runtime.Gosched()
 	}
+}
 
-	// Remove previous reader, if one exists.
-	if l.reader != nil {
-		_ = l.reader.Close()
-	}
+// stateLoop runs in a separate goroutine and runs the appropriate state loop.
+func (l *Log) stateLoop(closing <-chan struct{}, state State) {
+	defer l.wg.Done()
+	for {
+		// Transition to new state.
+		l.Logger.Printf("log state change: %s => %s", l.state, state)
+		l.state = state
 
-	// Set the new state.
-	l.state = state
+		// Remove previous reader, if one exists.
+		if l.reader != nil {
+			_ = l.reader.Close()
+		}
 
-	// Execute event loop for the given state.
-	switch state {
-	case Follower:
-		l.ch = make(chan struct{})
-		go l.followerLoop(l.ch)
-	case Candidate:
-		l.ch = make(chan struct{})
-		go l.candidateLoop(l.ch)
-	case Leader:
-		l.ch = make(chan struct{})
-		go l.leaderLoop(l.ch)
+		// Execute the appropriate state loop.
+		// Each loop returns the next state to transition to.
+		switch state {
+		case Stopped:
+			return
+		case Follower:
+			state = l.followerLoop(closing)
+		case Candidate:
+			state = l.candidateLoop(closing)
+		case Leader:
+			state = l.leaderLoop(closing)
+		}
 	}
 }
 
-// followerLoop continually attempts to stream the log from the current leader.
-func (l *Log) followerLoop(done chan struct{}) {
+func (l *Log) followerLoop(closing <-chan struct{}) State {
 	l.tracef("followerLoop")
-	var rch chan struct{}
+	defer l.tracef("followerLoop: exit")
+
+	// Ensure all follower goroutines complete before transitioning.
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	var transitioning = make(chan struct{})
+	defer close(transitioning)
+
+	// Read log from leader in a separate goroutine.
+	wg.Add(1)
+	go l.readFromLeader(&wg, transitioning)
+
 	for {
-		// Retrieve the term, last index, & leader URL.
-		l.mu.Lock()
-		if err := check(done); err != nil {
+		select {
+		case <-closing:
+			return Stopped
+		case <-l.initialize:
+			return Leader
+		case ch := <-l.Clock.AfterElectionTimeout():
+			l.closeReader()
+			close(ch)
+			return Candidate
+		case hb := <-l.heartbeats:
+			l.tracef("followerLoop: heartbeat: term=%d, idx=%d", hb.term, hb.commitIndex)
+
+			// Update term, commit index & leader.
+			l.mu.Lock()
+			if hb.term > l.term {
+				l.term = hb.term
+				l.votedFor = 0
+			}
+			if hb.commitIndex > l.commitIndex {
```

Comment:
Just so I am sure, here the index of the log is push forward to the index received over the heartbeat. I presume the entries corresponding to the greater `commitIndex` are also inserted into the log.